src/ClientHandle.cpp

```diff
@@ -184,7 +188,23 @@ void cClientHandle::Destroy(void)
 		{
 			player->StopEveryoneFromTargetingMe();
 			player->SetIsTicking(false);
-			world->RemovePlayer(player, true);
+
+			if (WasAddedToWorld)
+			{
+				// If ownership was transferred, our own smart pointer should be unset
+				ASSERT(!m_PlayerPtr);
+
+				m_PlayerPtr = world->RemovePlayer(*player, true);
+
+				// And RemovePlayer should have returned a valid smart pointer
+				ASSERT(m_PlayerPtr);
+			}
+			else
+			{
+				// If ownership was not transferred, our own smart pointer should be valid and RemovePlayer's should not
+				ASSERT(m_PlayerPtr);
+				ASSERT(!world->RemovePlayer(*player, true));
```

Comment:
I think the intention was that here, the player shouldn't have been added to the world at all (WasAddedToWorld is false, disconnection before Authenticate called) and so the debug-only call is just to verify that, at least for the appropriate chunk, it did not store any owning handles.

The threading relation between Authenticate and Destroy should be OK as you added a mutex guarding actions depending on the value of `m_State`.